src/invariant/InvariantTests.cpp

```diff
@@ -39,9 +53,494 @@ class TestInvariant : public Invariant
+    {
+        if (!mDeleted)
+        {
+            long long affectedRows = 0;
+            {
+                auto table = getTableName(mType);
+                auto& db = mApp.getDatabase();
+                auto prep = db.getPreparedStatement(
+                    "DELETE FROM " + table + " WHERE lastmodified= :v1 LIMIT 1");
+                auto& st = prep.statement();
+                st.exchange(soci::use(mFromLedgerSeq));
+                st.define_and_bind();
+                st.execute(true);
+                affectedRows = st.get_affected_rows();
+            }
+            assert(affectedRows <= 1);
+            if (affectedRows == -1)
+            {
+                throw std::runtime_error("Database operation unsupported");
+            }
+            mDeleted = (affectedRows == 1);
+        }
+        return ApplyBucketsWork::onSuccess();
+    }
+};
+
+class ApplyBucketsWorkModifyEntry : public ApplyBucketsWork
+{
+  private:
+    uint32_t mFromLedgerSeq;
+    LedgerEntryType mType;
+    bool mModified;
+
+    void maybeModifyAccountEntry()
+    {
+        auto& db = mApp.getDatabase();
+        std::string account;
+        {
+            auto prep = db.getPreparedStatement(
+                "SELECT accountid FROM accounts"
+                " WHERE lastmodified= :v1 LIMIT 1");
+            auto& st = prep.statement();
+            st.exchange(soci::into(account));
+            st.exchange(soci::use(mFromLedgerSeq));
+            st.define_and_bind();
+            if (!st.execute(true))
+            {
+                return;
+            }
+            assert(!st.fetch());
+        }
+
+        LedgerEntry entry;
+        entry.data.type(ACCOUNT);
+        entry.lastModifiedLedgerSeq = mFromLedgerSeq;
+        entry.data.account() = LedgerTestUtils::generateValidAccountEntry(5);
+        entry.data.account().accountID = KeyUtils::fromStrKey<PublicKey>(account);
+
+        LedgerHeader lh;
+        LedgerDelta ld(lh, db, false);
+        AccountFrame::pointer af = std::make_shared<AccountFrame>(entry);
+        af->storeChange(ld, db);
+        mModified = true;
+    }
+
+    void maybeModifyTrustLineEntry()
+    {
+        auto& db = mApp.getDatabase();
+        std::string account, issuer, assetCode;
+        uint32_t assetType;
+        {
+            auto prep = db.getPreparedStatement(
+                "SELECT accountid, issuer, assetcode, assettype FROM trustlines"
+                " WHERE lastmodified= :v1 LIMIT 1");
+            auto& st = prep.statement();
+            st.exchange(soci::into(account));
+            st.exchange(soci::into(issuer));
+            st.exchange(soci::into(assetCode));
+            st.exchange(soci::into(assetType));
+            st.exchange(soci::use(mFromLedgerSeq));
+            st.define_and_bind();
+            if (!st.execute(true))
+            {
+                return;
+            }
+            assert(!st.fetch());
+        }
+
+        LedgerEntry entry;
```

Comment:
this code goes away if you use the regular load method, right?

Reply:
The same comments as above apply to replacing the above SQL query with the regular load method. If I had a key for an entry with the appropriate ledger number (either through the above SQL query or through maintaining that information internally) and used the key to invoke the regular load method, this code would be simplified significantly. It is not so easy unfortunately. TrustFrame::loadTrustLine takes an Asset as a parameter. But to construct the object of type Asset (defined in XDR), would require much of this exact code specifically the if (assetType == ...) part. In net, little can be gained in terms of removing this code by using the regular load functions.